planner/core/exhaust_physical_plans.go

```diff
@@ -2227,6 +2227,65 @@ func (la *LogicalAggregation) getStreamAggs(prop *property.PhysicalProperty) []P
 	return streamAggs
 }
 
+// TODO: support more operators and distinct later
+func (la *LogicalAggregation) checkCanPushDownToMPP() bool {
+	for _, agg := range la.AggFuncs {
+		if agg.Name != ast.AggFuncSum && agg.Name != ast.AggFuncMin && agg.Name != ast.AggFuncCount && agg.Name != ast.AggFuncMax && agg.Name != ast.AggFuncFirstRow {
+			return false
+		}
+		if agg.HasDistinct {
+			return false
+		}
+	}
+	return true
+}
+
+func (la *LogicalAggregation) getHashAggsForMPP(prop *property.PhysicalProperty) (hashAggs []PhysicalPlan) {
+	groupByCols := la.GetGroupByCols()
+	if prop.PartitionTp == property.HashType {
+		if matches := prop.IsSubsetOf(groupByCols); len(matches) != 0 {
+			matchedCols := chooseSubsetOfJoinKeys(groupByCols, matches)
+			// 1-phase
+			childProp := &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: math.MaxFloat64, PartitionTp: property.HashType, PartitionCols: matchedCols, CanAddEnforcer: true}
+			agg := NewPhysicalHashAgg(la, la.stats.ScaleByExpectCnt(prop.ExpectedCnt), childProp)
+			agg.SetSchema(la.schema.Clone())
+			agg.MppRunMode = Mpp1Phase
+			hashAggs = append(hashAggs, agg)
+			// 2-phase
+			childProp = &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: math.MaxFloat64, PartitionTp: property.AnyType}
+			agg = NewPhysicalHashAgg(la, la.stats.ScaleByExpectCnt(prop.ExpectedCnt), childProp)
+			agg.SetSchema(la.schema.Clone())
+			agg.MppRunMode = Mpp2Phase
+			agg.PartitionCols = matchedCols
+			hashAggs = append(hashAggs, agg)
+		}
+	} else {
+		if len(groupByCols) != 0 {
+			// TODO: permute various partition columns from group-by columns
+			// 1-phase
+			childProp := &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: math.MaxFloat64, PartitionTp: property.HashType, PartitionCols: groupByCols, CanAddEnforcer: true}
+			agg := NewPhysicalHashAgg(la, la.stats.ScaleByExpectCnt(prop.ExpectedCnt), childProp)
+			agg.SetSchema(la.schema.Clone())
+			agg.MppRunMode = Mpp1Phase
+			hashAggs = append(hashAggs, agg)
+			// 2-phase
+			childProp = &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: math.MaxFloat64, PartitionTp: property.AnyType}
+			agg = NewPhysicalHashAgg(la, la.stats.ScaleByExpectCnt(prop.ExpectedCnt), childProp)
+			agg.SetSchema(la.schema.Clone())
+			agg.MppRunMode = Mpp2Phase
+			agg.PartitionCols = groupByCols
+			hashAggs = append(hashAggs, agg)
+		}
+		// TODO: support running scalar agg in MPP, merge the final result to one node
```

Comment:
As I have mentioned, a group-by agg can also do a TiDB-TiFlash agg, in case that the father requires a root task and the cardinality of gby columns is small.